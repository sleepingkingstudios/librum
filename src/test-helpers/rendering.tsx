import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
