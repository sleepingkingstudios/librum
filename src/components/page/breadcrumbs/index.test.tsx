import * as React from 'react';

import '@testing-library/jest-dom';

import { PageBreadcrumbs } from './index';
import type { Breadcrumb } from './types';
import { render } from '@test-helpers/rendering';

describe('<PageBreadcrumbs />', () => {
  describe('with breadcrumbs: an empty array', () => {
    const breadcrumbs: Breadcrumb[] = [];

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PageBreadcrumbs breadcrumbs={breadcrumbs} />,
        { router: true }
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with breadcrumbs: an array of breadcrumbs', () => {
    const breadcrumbs: Breadcrumb[] = [
      {
        label: 'Home',
        url: '/',
      },
      {
        label: 'Admin'
      },
      {
        label: 'Example',
        url: '/example',
      },
    ];

    it('should match the snapshot', () => {
      const { asFragment, getByText } = render(
        <PageBreadcrumbs breadcrumbs={breadcrumbs} />,
        { router: true }
      );

      breadcrumbs.forEach((breadcrumb: Breadcrumb) => {
        const { label } = breadcrumb;

        expect(getByText(label)).toBeVisible();
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
