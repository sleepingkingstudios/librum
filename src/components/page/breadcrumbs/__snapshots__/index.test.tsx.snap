// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageBreadcrumbs /> with breadcrumbs: an array of breadcrumbs should match the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col sm:flex-row page-breadcrumbs"
  >
    <li
      class="page-breadcrumb"
    >
      <span
        class="page-breadcrumb-empty"
      >
        Home
      </span>
    </li>
    <li
      class="page-breadcrumb"
    >
      <span
        class="page-breadcrumb-empty"
      >
        Admin
      </span>
    </li>
    <li
      class="page-breadcrumb"
    >
      <a
        class="page-breadcrumb-link"
        href="/example"
      >
        Example
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<PageBreadcrumbs /> with breadcrumbs: an empty array should match the snapshot 1`] = `<DocumentFragment />`;
